Better backup and restore

garrettr opened the pull request:
This improves the backup and restore Ansible roles. The primary goals are to make backups easier, faster, and less error-prone.

## Key Changes

* _Greatly_ simplifies the backup and restore scripts.
  * **Important Note**: this happens at the expense of flexibility. The new scripts **assume** that both the source and destination installations use the default paths for SecureDrop and Tor. Based on my experience, this is a reasonable assumption, and it is documented in comments on the scripts.
* Stops encrypting the backup files to the air-gapped key.
  * This does not provide a security benefit. All of the sensitive files (submissions) in the backup are already encrypted to the air-gapped key. Also, the Admin Workstation has full access (via SSH) to the App Server, so it can be trusted with data from the App Server and there is no need for an additional layer of encryption. 
  * Not encrypting the backup file makes restoring from backup *much* less complicated and error-prone.
* Stops "secure deleting" the backup files.
  * No security benefit.
  * Creates problems for submissions that have a non-trivial amount of submissions (e.g. > 100MB). Secure deleting this amount of data takes a *very* long time, which creates all kinds of problems.
  * This is a problem for systems that use hardware RAID, since `shred` and hardware RAID do not play nice.
* Renames the variable that needs to be set to trigger a restore, from `backup_zip` to `restore_file`, which is much clearer. In the past, some admins were confused and (reasonably) thought that `backup_zip` was the name of the file to backup *to*, rather than the file to restore *from*.

## Testing

- [x] Since this PR modifies `prod-specific.yml`, test: one should be able to `git pull` and then cleanly re-apply an existing production configuration, using the stash-pull-pop workflow.
- [x] Test that this handles large amounts of submissions well (both backup and restore)
  - [x] Backup large SeucreDrop over Tor
  - [x] Restore large SecureDrop over Tor
- [x] Ensure that restoring SSH ATHS does not cause problems for the admin.
 - [x] Does it prevent the restore play from completing?
 - [x] What is the best way to sync the Admin Workstation with the restored server state?

ageis commented on install_files/ansible-base/roles/backup/files/0.3_restore.py:
```diff
@@ -1,155 +1,50 @@
 #!/usr/bin/python2.7
 """
+This script and backup archive should be copied to the App server and run by
+the Anisble playbook. When run (as root), it restores the contents of the 0.3
```
spell check on Anisble

garrettr commented:
Currently uploading several GB of data to a test SecureDrop instance so I can do a realistic test of backup and restore (i.e. to compare to a production server that may have been running for years and accumulated a significant number of submissions).

garrettr commented:
Verified that the stash-pull-pop workflow works with the changes to `prod-specific.yml` in this PR. On a test production machine, Git was able to automerge the file and no manual merge conflict resolution was required.

garrettr commented:
Verified migration by backing up the SecureDrop Test Instance with 2GB of submissions and then restoring it onto some staging VM's. It would be good to test the restore over Tor to make sure it's unlikely to time out.

conorsch commented on install_files/ansible-base/roles/backup/tasks/main.yml:
```diff
@@ -1,50 +1,6 @@
-  when: backup_zip is defined
+  # Perform a restoration if a restore file is specified
+- include: restore.yml
+  when: restore_file is defined
```
Nit: `when: restore_file is defined and restore_file != ''`. An empty string should be treated the same as undefined.

garrettr replied:
> An empty string should be treated the same as undefined.

Really? Obviously an empty string is an invalid value here, but if the variable is defined at all then somebody has modified the configuration to define it (even if they did so incorrectly). I am inclined to say "if you defined a variable, you're responsible for defining it correctly" (at least for now).

garrettr replied:
To clarify my stance on saying "if you defined a variable, you're responsible for defining it correctly": we can't comprehensively verify whether the values an admin sets for variables like `restore_file` are correct, so IMHO we should not bother doing a half-assed job (e.g. checking for one obviously invalid but also unlikely value, like `''`).

At some point we will overhaul the configuration mechanism for SecureDrop, and we'll probably add a layer of abstraction between the configuration settings that the admin touches versus the `.yml` that gets fed to Ansible (or whatever). I think we should leave this type of concern for that more comprehensive overhaul of SecureDrop configuration management, instead of doing a cursory job now.

garrettr replied:
From in-person review, @conorsch agrees this is fine the way it is.

conorsch commented on install_files/ansible-base/roles/backup/tasks/restore.yml:
```diff
@@ -0,0 +1,15 @@
+---
+- name: copy restore script to app server
+  copy: src=0.3_restore.py dest=/tmp/ owner=root mode=770
```
For consistency, use the multiline YAML syntax here, as in the `file:` task below:

```
- name: Copy restore file to App Server.
  copy:
    src: "{{ restore_file }}"
    dest: /tmp/{{ restore_file }}
```

conorsch commented on install_files/ansible-base/staging-specific.yml:
```diff
@@ -46,11 +46,3 @@ sasl_username: "test"
-# ex: backup_zip: "sd-backup-2015-01-15--21-03-32.zip"
-# To use uncomment the following line and enter the correct decrypted zip
-# filename between the quotes.
-#backup_zip: ""
```
Skipping the backup action by default makes sense for staging. :+1:

garrettr replied:
@conorsch Yeah. Obviously all of this config management needs an overhaul anyway but this seemed like a reasonable short-term change.

garrettr replied:
Note that this doesn't "skip the backup action" as @conorsch says. The backup role always runs. This simply removes an outdated comment, because this PR replaces `backup_zip` for the more descriptive variable `restore_file`.

conorsch commented:
> On a test production machine, Git was able to automerge the file and no manual merge conflict resolution was required.

That's a pleasant surprise!

Backporting these changes onto `develop` will require manual conflict resolution, however, since the `develop` branch already has multiline YAML syntax implemented on all tasks for this role, but without the new functionality.

garrettr commented:
Testing a large restore over Tor right now. To do so, I:

* Copied the `sd-backup-*.tar.gz` backup file from the test server to my development machine.
* `vagrant up /staging$/` up a staging environment
 * I could've used `/prod$/`, but had already set up staging and didn't want to wait for a rebuild.
* Modified Vagrant for the staging environment so vagrant+ansible would do SSH over Tor, by cherry-picking the relevant changes for `SECUREDROP_SSH_OVER_TOR` from the develop branch.
 * Would not have had to do this if I had used `prod`, since `prod` fully emulates a production installation and always does SSH over Tor.
* Added `ansible.tags = ['backup']` to Vagrantfile so provisioning only runs the backup tag. This is to save time, since every Ansible command is happening over Tor now.
* `SECUREDROP_SSH_OVER_TOR=1 vagrant provision app-staging`
* ...wait...

garrettr commented:
Confirmed that a large restore over Tor eventually succeeded, although it took a very long time (2GB restore took ~5 hours).

garrettr commented:
Ansible also appeared to continue running successfully even after the restore script replaced the HidServAuth values and reloaded Tor, which is convenient (if a little surprising): [Ansible logs](https://gist.github.com/garrettr/1f808770d692f51b1e1a0059c336df5f).

garrettr commented:
> Ensure that restoring SSH ATHS does not cause problems for the admin.

Let's expand on this a little bit.

The data that gets backed up and restored from a SecureDrop includes the hidden service configuration information (from `/var/lib/tor/services`). This allows a restored SecureDrop to maintain the same .onion addresses for the Source Interface, Document Interface, and SSH interfaces.

The sequence of events for a backup and restore is as follows:

1. A SecureDrop was installed and has been running for some time.
2. An admin runs the `backup` role on the installed SecureDrop, creating a snapshot of that SecureDrop's state which is stored in an archive file.
3. A new SecureDrop is installation is performed. New hidden services are automatically generated.
4. In order to connect to the new installation to restore the backup onto it, the admin must update the Ansible inventory and configure Tor to connect to the SSH ATHS, which is done via `tails_files/install.sh`.
5. The admin runs the restore task, connecting to the new SecureDrop installation using the *new* ATHS credentials.
6. The restore role copies the **old** values over the new values, and reloads Tor.

Concerns to investigate:

- [ ] Does restoring the old SSH ATHS values break the Ansible session running the restore play?
- [ ] Once the restore role is done running, the files on the Admin Workstation (`ansible_base/*ths`, Ansible inventory, tails_files torrc_additions, etc.) will out of sync in relation to the restored state on the server. What is the best way to sync the Admin Workstation with the restored server state?

garrettr commented:
Adding support-request since this work was initially motivated by multiple requests from SecureDrop admins who want to move their systems to different hardware. The "backup and restore" Ansible roles are the easiest way to do that, but they needed a bunch of work, which is described in detail in the top comment.

garrettr commented:
>  Does restoring the old SSH ATHS values break the Ansible session running the restore play?

@fowlslegs investigated this in https://github.com/freedomofpress/securedrop/pull/1293#issue-153916849 and convincingly argued that, because reloading Tor marks circuits as dirty but *does not kill them*, restoring the SSH ATHS values will not break the currently running Ansible session *as long as it continues to use the same TCP connection*.

We noted that this may only work because we use `ControlPersist` in our SSH configuration for Ansible (https://github.com/freedomofpress/securedrop/pull/1293#issuecomment-221116353). It's important to test this thoroughly, and keep in mind that changing the Ansible SSH configuration could have a negative impact on the restore role.

garrettr commented:
> Once the restore role is done running, the files on the Admin Workstation (ansible_base/*ths, Ansible inventory, tails_files torrc_additions, etc.) will be out of sync in relation to the restored state on the server. What is the best way to sync the Admin Workstation with the restored server state?

In https://github.com/freedomofpress/securedrop/pull/1273/commits/cecd7e54865fca2a4cf5244ae1104f450a5dfb0c, I swapped the order of the restore role and the restrict_direct_access_app role, so when restrict_direct_access_app fetchs the Tor ATHS credentials, it does so _after_ the restore has been completed, which should keep the Admin Workstation in sync with the Tor ATHS credentials that are on the server after a restore.

garrettr commented:
Currently re-testing backup and restore to ensure:

1. Re-ordering the backup/restore and restrict_direct_access_app roles doesn't lead to problems with losing the Tor connection. With the roles swapped, the number of Ansible tasks that have to happen over a now-"dirty" connection is increased, and this might lead to problems.
2. State on the "Admin Workstation" is appropriately synchronized after performing a restore to a fresh installation.

garrettr commented:
After discussing the decision with @conorsch, I added the `backup` tag to the `{app,mon}_display_onions` tasks, which are responsible for fetching the ATHS credentials from the servers to the Admin Workstation. This allows an admin to run a backup and/or restore operation in isolation using `ansible-playbook install_files/ansible-base/securedrop-prod.yml --tags=backup`, which is faster, more convenient, and will make debugging easier in case something goes wrong with the backup and/or restore process.

garrettr commented:
I have fully tested backing up and restoring a small instance (~50MB restore file) in a variety of contexts using the production Vagrant environment. Everything looks good! I am going to kick off a test of restoring a 2GB restore file later today, the results of that should be available by tomorrow.

[Detailed testing notes](https://www.evernote.com/l/AAiIUyoi8jVPArVwE7XyUIN_-AUhUhvvL2U).

garrettr commented:
Tried to test a 2GB restore overnight last night, but it failed in the middle of the night. My computer probably tried to go to sleep during the transfer of the restore file. Trying it again now, and taking notes in the same Evernote from yesterday.

garrettr commented:
2GB restore succeeded! Detailed notes in the Evernote. Unfortunately, since this is a restore of the hardware test instance in our other office, I need to wait until tomorrow so I can shut down the test hardware instance in order to properly test the restore.

garrettr commented:
Rebased on top of `release/0.3.7` so I could write documentation for the new-and-improved backup and restore procedures for Sphinx/ReadTheDocs, which needed the backported ReadTheDocs documentation added to `release/0.3.7` in b39bf3868e2fe8bad8d5813951893068e239f85e. Had to force push the feature branch as a result of the rebase.

@conorsch I'd appreciate a review of the backup and restore documentation added in 90908aa3c9bf974525e17e5f0583b4c3e3ff04f9. You can pull it locally and build the documentation with `make html` or `sphinx auto-build`. For your convenience, I also published a copy on ReadTheDocs here: https://docs.securedrop.org/en/improve-backup-and-restore/backup_and_restore.html.

conorsch commented:
@garrettr Reviewed docs, pushed up some revisions. Mostly the changes strive for concision. There are a few niggling details that I'd like to improve about the docs, but are out of scope for this PR:

* SSH trouble shooting steps belong in **Test the installation**, to make docs DRY
* we're still requiring Admins to provide `-i inventory` on the command line when we can store that in the `ansible.cfg` file and save them the trouble

Give my edits a sanity check, but with their inclusion I'm happy with merging. :+1: Thanks for testing so extensively!

garrettr commented:
> There are a few niggling details that I'd like to improve about the docs, but are out of scope for this PR

Agreed; let's file follow-up issues for these. Good ideas, but we should stay on target to get 0.3.7 released ASAP.

conorsch commented:
> Good ideas, but we should stay on target to get 0.3.7 released ASAP.

:+1: :tada: :shipit: :shipit: :shipit:

garrettr commented:
> Give my edits a sanity check, but with their inclusion I'm happy with merging.

Sanity checked, lgtm. I added a note about how you may have to wait between deleting submissions on the Document Interface and actually reclaiming space in `/var/lib/securedrop/store` (due to file deletion being asynchronous, queued, and slow due the use of `srm`).

This is ready to merge.

garrettr commented:
> This is ready to merge.

Hold up. There's one more piece that @conorsch and I discussed this week that still needs some thought. Currently, the backup role runs by default whenever the full playbook is run. This could be problematic for production installations that have a lot of submissions.

If they don't plan ahead for the backup (in terms of available disk space on the Admin Workstation, time needed to transfer the backup archive over Tor, etc.), this could be a bad experience for them if they need to run a playbook for a reason other than backup/restore - e.g. we ask them to run the playbook for an upgrade that requires Ansible, such as 0.3.6).

I think we should consider _not_ running the backup role by default, so it _only_ gets run when the admin explicitly chooses to do so (following the new documentation introduced in 90908aa). @conorsch What do you think?

conorsch commented:
> I think we should consider _not_ running the backup role by default, so it _only_ gets run when the admin explicitly chooses to do so

Agreed, that's logical behavior, and will be particularly useful for long-running instances where too-frequent backups can clutter up the Admin Workstation. Worse, the long-running backups may discourage Admins from running the playbooks frequently.

The ideal solution to achieve the only-when-requested backup behavior is an isolated playbook encapsulating the backup and restore logic. That addition is best deferred until 0.4, however, due to substantial changes from the layout of the playbooks and roles that has already occurred in develop. As an intermediate solution, we can disable the backup role by default at the play level within `securedrop-prod.yml`, and reenable it conditionally if either `perform_backup=true` or `restore_file` is defined. 

14cd246 implements the intermediate solution, and tweaks the docs for upgrade accordingly. This should be tested on hardware again.

garrettr commented:
Awesome. I think this is finally done!!

conorsch commented:
Tracking a list of issues that will be resolved once this PR is merged:

* closes  #942
* closes #1127 
* closes #1263 

There are a few "related" issues that have already been linked via discussion. Not marking those for automatic close upon merge.
